.github/workflows/ci.yml

```diff
@@ -24,6 +24,10 @@ jobs:
   codegen:
     name: codegen
     runs-on: ubuntu-20.04
+    # https://docs.github.com/en/actions/using-jobs/using-concurrency
+    concurrency:
+      group: codegen-${{ github.workflow }}-${{ github.head_ref || github.run_id }}
```

Comment:
in what situation value ${{ github.run_id }} will be selected?

Reply:
https://docs.github.com/en/actions/using-jobs/using-concurrency#example-using-a-fallback-value

The `github.head_ref` is only defined on pull_request events. if `github.head_ref` is undefined, the concurrency group will fallback to the `github.run_id`, which is guaranteed to be both unique and defined for the run.